File: dotnet/src/OrderWorkflowScenarios.cs
using System.Diagnostics;
using DotNetOrderManagement.model;
using Microsoft.Extensions.Logging;
using System.Threading.Tasks;

using Temporalio.Common;
using Temporalio.Converters;
using Temporalio.Exceptions;
using Temporalio.Workflows;

namespace DotNetOrderManagement;

[Workflow(Dynamic = true)]
public class OrderWorkflowScenarios
{
    private static readonly string BUG = "OrderWorkflowRecoverableFailure";
    private static readonly string CHILD = "OrderWorkflowChildWorkflow";
    private static readonly string SIGNAL = "OrderWorkflowHumanInLoopSignal";
    private static readonly string UPDATE = "OrderWorkflowHumanInLoopUpdate";
    private static readonly string VISIBILITY = "OrderWorkflowAdvancedVisibility";
    
    private SearchAttributeKey<string> ORDER_STATUS_SA = SearchAttributeKey.CreateKeyword("OrderStatus");

    private int _progress = 0;
    private string _updatedAddress = string.Empty;

    [WorkflowRun]
    public async Task<OrderOutput> RunAsync(IRawValue[] args)
    {
        var logger = Workflow.Logger;
        var type = Workflow.Info.WorkflowType;
        var input = Workflow.PayloadConverter.ToValue<OrderInput>(args[0]);
        logger.LogInformation("Dynamic Order workflow started, type = {}, orderId = {}", type, input.OrderId);
        
        // Saga compensations
        var compensations = new List<Func<Task>>();

        // Get Items
        var orderItems = await Workflow.ExecuteLocalActivityAsync((OrderActivities act) => 
            act.GetItems(), OrderActivities.LocalActivityOpts);

        await UpdateProgress("Check Fraud", 0, 0);
        
        // Check fraud
        await Workflow.ExecuteActivityAsync((OrderActivities act) =>
            act.CheckFraud(input), OrderActivities.ActivityOpts);
        
        await UpdateProgress("Prepare Shipment", 25, 1);
        
        // Prepare shipment
        compensations.Add(async () => 
            await Workflow.ExecuteActivityAsync(() => 
                OrderActivities.UndoPrepareShipment(input), OrderActivities.ActivityOpts));
        
        await Workflow.ExecuteActivityAsync((OrderActivities act) => 
            act.PrepareShipment(input), OrderActivities.ActivityOpts);
        
        await UpdateProgress("Charge Customer", 50, 1);
        
        // Charge Customer
        try
        {
            compensations.Add(async () => 
                await Workflow.ExecuteActivityAsync(() => 
                    OrderActivities.UndoChargeCustomer(input), OrderActivities.ActivityOpts));
            
            await Workflow.ExecuteActivityAsync((OrderActivities act) =>
                act.ChargeCustomer(input, type), OrderActivities.ActivityOpts);
        }
        catch (ActivityFailureException af)
        {
            logger.LogError("Failed to charge customer {}", af);
            await SagaCompensate(compensations);
            throw;
        }
        
        await UpdateProgress("Ship Order", 75, 3);

        if (BUG.Equals(type))
        {
            // Simulate a bug
            // throw new Exception("Simulated bug - fix me!");
        }

        if (SIGNAL.Equals(type) || UPDATE.Equals(type))
        {
            // await message to update address
            await WaitForUpdatedAddressOrTimeout(input);
        }

        // Ship order items
        if (CHILD.Equals(type))
        {
            await ShipOrderChildWorkflowsAndWait(input, orderItems, type);
        }
        else
        {
            await ShipOrderActivitiesAndWait(input, orderItems, type);
        }
        logger.LogInformation("Items have been shipped");
        await UpdateProgress("Order Completed", 100, 2);    
        
        // Generate tracking ID
        var trackingId = Workflow.Random.Next().ToString();
        return new OrderOutput(trackingId, input.Address);
    }
    
    [WorkflowQuery("getProgress")]
    public int QueryProgress()
    {
        return _progress;
    }
    
    [WorkflowSignal("UpdateOrder")]
    public async Task UpdateOrderSignal(UpdateOrderInput updateInput)
    {
        Workflow.Logger.LogInformation("Received update order signal with address: {}", updateInput.Address);
        _updatedAddress = updateInput.Address;
    }

    [WorkflowUpdate("UpdateOrder")]
    public async Task<string> UpdateOrderUpdate(UpdateOrderInput updateInput)
    {
        Workflow.Logger.LogInformation("Received update order update with address: {}", updateInput.Address);
        _updatedAddress = updateInput.Address;
        return "Updated address: " +_updatedAddress;
    }

    [WorkflowUpdateValidator("UpdateOrderUpdate")]
    public void UpdateOrderValidator(UpdateOrderInput updateInput)
    {
        if (!Char.IsDigit(updateInput.Address[0]))
        {
            Workflow.Logger.LogInformation("Rejecting order update, invalid address: {}", updateInput.Address);
            throw new ApplicationFailureException("Address must start with a digit", "invalid-address");
        }
        
        Workflow.Logger.LogInformation("Order update address is valid: {}", updateInput.Address);
    }

    private async Task WaitForUpdatedAddressOrTimeout(OrderInput input)
    {
        Workflow.Logger.LogInformation("Waiting up to 60 seconds for updated address");
        var ok = await Workflow.WaitConditionAsync(
            () =>  !string.IsNullOrEmpty(_updatedAddress), TimeSpan.FromSeconds(60));
        if (ok)
        {
            input.Address = _updatedAddress;
        }
        else
        {
            // Do nothing - use the original address
            // In other cases, you may want to throw an exception on timeout, e.g.
            // throw new ApplicationFailureException("Updated address was not received in 60 seconds");
        }
        
    }
    
    private async Task ShipOrderActivitiesAndWait(OrderInput input, List<OrderItem> items, string type)
    {
        var logger = Workflow.Logger;
        var activities = new List<Task>();
        foreach (var orderItem in items)
        {
            logger.LogInformation("Shipping item: {}", orderItem.Description);
            activities.Add(
                Workflow.ExecuteActivityAsync((OrderActivities act) => 
                    act.ShipOrder(input, orderItem), OrderActivities.ActivityOpts)
            );
        }
        // Wait for all items to ship
        logger.LogInformation("Waiting for shipping activities to complete...");
        await Workflow.WhenAllAsync(activities);
        logger.LogInformation("Shipping activities are now completed");
    }
    
    private async Task ShipOrderChildWorkflowsAndWait(OrderInput input, List<OrderItem> items, string type)
    {
        Debug.Assert(CHILD.Equals(type));   
        var logger = Workflow.Logger;
    
        var childWorkflows = new List<Task>();
        foreach (var orderItem in items)
        {
            logger.LogInformation("Shipping item via child workflow: {}, ", orderItem.Id);
            var handle = await Workflow.StartChildWorkflowAsync(
                (ShippingChildWorkflow wf) => wf.Execute(input, orderItem), new()
                {
                    Id = string.Join("-", "shipment", input.OrderId, orderItem.Id),
                    ParentClosePolicy = ParentClosePolicy.Terminate,
                });
            
            childWorkflows.Add(handle.GetResultAsync());
        }
    
        logger.LogInformation("Waiting for child shipping workflows to complete...");
        await Workflow.WhenAllAsync(childWorkflows);
        logger.LogInformation("Shipping child workflows are now completed");
    }
    
    private async Task SagaCompensate(List<Func<Task>> compensations)
    {
        compensations.Reverse(0, compensations.Count);
        foreach (var comp in compensations)
        {
#pragma warning disable CA1031
            try
            {
                await comp.Invoke();
            }
            catch (Exception ex)
            {
                Workflow.Logger.LogError(ex, "Failed to compensate");
                // don't propagate the exception
            }
        }
#pragma warning restore CA1031
    }

    private async Task UpdateProgress(string orderStatus, int progress, int sleep)
    {
        _progress = progress;
        if (sleep > 0)
        {
            await Workflow.DelayAsync(TimeSpan.FromSeconds(sleep));
        }

        if (VISIBILITY.Equals(Workflow.Info.WorkflowType))
        {
            Workflow.Logger.LogInformation("Advanced visibility .. {}", orderStatus);
            Workflow.UpsertTypedSearchAttributes(ORDER_STATUS_SA.ValueSet(orderStatus));    
        }
    }

}

File: python/order_workflow_scenarios.py
import logging
from datetime import timedelta
import asyncio
from collections.abc import Sequence
from typing import Any

from temporalio import workflow
from temporalio.common import RawValue, SearchAttributeKey
from temporalio.exceptions import ApplicationError
from temporalio.workflow import ParentClosePolicy

from activities import OrderActivities
from shared_objects import OrderInput, OrderOutput, UpdateOrderInput, OrderItem
from shipping_child_workflow import ShippingChildWorkflow

logging.basicConfig(level=logging.INFO)


@workflow.defn(dynamic=True)
class OrderWorkflowScenarios:

    BUG = "OrderWorkflowRecoverableFailure"
    CHILD = "OrderWorkflowChildWorkflow"
    SIGNAL = "OrderWorkflowHumanInLoopSignal"
    UPDATE = "OrderWorkflowHumanInLoopUpdate"
    VISIBILITY = "OrderWorkflowAdvancedVisibility"

    ORDER_STATUS_SA = SearchAttributeKey.for_keyword("OrderStatus")

    def __init__(self) -> None:
        self.progress = 0
        self.updated_address = None
        self.retry_policy = OrderActivities.retry_policy

    @workflow.run
    async def execute(self, args: Sequence[RawValue]) -> Any:
        input = workflow.payload_converter().from_payload(args[0].payload, OrderInput)
        workflow_type = workflow.info().workflow_type
        workflow.logger.info(f"Dynamic Order workflow started, type = {workflow_type}, orderId = {input.OrderId}")

        # Saga compensations
        compensations = []

        # Get items
        order_items = await workflow.execute_local_activity_method(
            OrderActivities.get_items,
            start_to_close_timeout=timedelta(seconds=5)
        )

        await self.update_progress("Check Fraud", 0, 0)

        # Check fraud
        await workflow.execute_activity_method(
            OrderActivities.check_fraud,
            input,
            start_to_close_timeout=timedelta(seconds=5),
            retry_policy=self.retry_policy
        )

        await self.update_progress("Prepare Shipment", 25, 1)

        # Prepare shipment
        compensations.append(OrderActivities.undo_prepare_shipment)
        await workflow.execute_activity_method(
            OrderActivities.prepare_shipment,
            input,
            start_to_close_timeout=timedelta(seconds=5),
            retry_policy=self.retry_policy
        )

        await self.update_progress("Charge Customer", 50, 1)

        # Charge customer
        try:
            compensations.append(OrderActivities.undo_charge_customer)
            await workflow.execute_activity_method(
                OrderActivities.charge_customer,
                args=[input, workflow_type],
                start_to_close_timeout=timedelta(seconds=5),
                retry_policy=self.retry_policy
            )
        except Exception as ex:
            workflow.logger.error("Failed to charge customer", ex)
            for compensation in reversed(compensations):
                await workflow.execute_activity(
                    compensation,
                    input,
                    start_to_close_timeout=timedelta(seconds=10),
                    retry_policy=self.retry_policy
                )
            raise ex

        await self.update_progress("Ship Order", 75, 3)

        if self.BUG == workflow_type:
            # Simulate bug
            raise RuntimeError("Simulated bug - fix me!")
            pass

        if self.SIGNAL == workflow_type or self.UPDATE == workflow_type:
            # Await message to update address
            await self.wait_for_updated_address_or_timeout(input)

        # Ship order items
        handles = []
        for item in order_items:
            workflow.logger.info(f"Shipping item: {item.description}")
            handles.append(self.ship_item_async(input, item, workflow_type))

        # Wait for all items to ship
        await asyncio.gather(*handles)

        await self.update_progress("Order Completed", 100, 0)

        # Generate trackingId
        tracking_id = str(workflow.uuid4())
        return OrderOutput(tracking_id, input.Address)

    def ship_item_async(self, input: OrderInput, item: OrderItem, workflow_type: str) -> Any:
        if self.CHILD == workflow_type:
            return workflow.execute_child_workflow(
                ShippingChildWorkflow.execute,
                args=[input, item],
                id="shipment-" + input.OrderId + "-" + str(item.id),
                parent_close_policy=ParentClosePolicy.TERMINATE
            )

        else:
            return workflow.execute_activity_method(
                OrderActivities.ship_order,
                args=[input, item],
                start_to_close_timeout=timedelta(seconds=5),
                retry_policy=self.retry_policy
            )

    async def wait_for_updated_address_or_timeout(self, input: OrderInput):
        workflow.logger.info("Waiting up to 60 seconds for updated address")
        try:
            await workflow.wait_condition(lambda: self.updated_address != None, timeout=timedelta(seconds=60))
            input.Address = self.updated_address
        except asyncio.TimeoutError:
            pass
            # raise ApplicationFailure("Updated address was not received within 60 seconds.", type="timeout")

    async def update_progress(self, order_status: str, progress: int, sleep: int):
        self.progress = progress
        if sleep > 0:
            await asyncio.sleep(sleep)
        if self.VISIBILITY == workflow.info().workflow_type:
            workflow.upsert_search_attributes([self.ORDER_STATUS_SA.value_set(order_status)])

    @workflow.query(name="getProgress")
    def query_progress(self) -> int:
        return self.progress

    @workflow.signal(name="UpdateOrder")
    def update_order_signal(self, update_input: UpdateOrderInput):
        workflow.logger.info(f"Received update order signal with address: {update_input.Address}")
        self.updated_address = update_input.Address

    @workflow.update(name="UpdateOrder")
    def update_order_update(self, update_input: UpdateOrderInput) -> str:
        workflow.logger.info(f"Received update order signal with address: {update_input.Address}")
        self.updated_address = update_input.Address
        return "Updated address: " + update_input.Address

    @update_order_update.validator
    def update_order_validator(self, update_input: UpdateOrderInput):
        if not update_input.Address[0].isdigit():
            workflow.logger.info(f"Rejecting order update, invalid address: {update_input.Address}")
            raise ApplicationError("Address must start with a digit", type="invalid-address")
        workflow.logger.info(f"Order update address is valid: {update_input.Address}")
